holoviews/element/sankey.py

```diff
@@ -0,0 +1,344 @@
+from __future__ import division
+
+from functools import cmp_to_key
+from itertools import cycle
+
+import param
+import numpy as np
+
+from ..core.dimension import Dimension
+from ..core.data import Dataset
+from ..core.operation import Operation
+from ..core.util import OrderedDict, unique_array, RecursionError
+from .graphs import Graph, redim_graph
+from .util import quadratic_bezier
+
+
+class layout_sankey(Operation):
+    """
+    Computes a Sankey diagram from a Graph element.
+
+    Adapted from d3-sankey under BSD-3 license.
+    """
+
+    bounds = param.NumericTuple(default=(0, 0, 1000, 500))
+
+    node_width = param.Number(default=15, doc="""
+        Width of the nodes.""")
+
+    node_padding = param.Integer(default=10, doc="""
+        Number of pixels of padding relative to the bounds.""")
+    
+    iterations = param.Integer(default=32, doc="""
+        Number of iterations to run the layout algorithm.""")
+
+    def _process(self, element, key=None):
+        graph = {'nodes': [], 'links': []}
+        self.computeNodeLinks(element, graph)
+        self.computeNodeValues(graph)
+        self.computeNodeDepths(graph)
+        self.computeNodeBreadths(graph)
+        self.computeLinkBreadths(graph)
+        paths = self.computePaths(graph)
+
+        node_data = []
+        for node in graph['nodes']:
+            node_data.append((np.mean([node['x0'], node['x1']]),
+                              np.mean([node['y0'], node['y1']]),
+                              node['index'])+tuple(node['values']))
+        if element.nodes.ndims == 3:
+            kdims = element.nodes.kdims
+        elif element.nodes.ndims:
+            kdims = element.node_type.kdims[:2] + element.nodes.kdims[-1:]
+        else:
+            kdims = element.node_type.kdims
+        nodes = element.node_type(node_data, kdims=kdims, vdims=element.nodes.vdims)
+        edges = element.edge_type(paths)
+        sankey = Sankey((element.data, nodes, edges), compute=False)
```

Comment:
It is rather odd how the ``Sankey`` element uses this operation in its constructor which then returns a ``Sankey`` object!  It would be better to have a method just computing the information needed *by* ``Sankey`` and using that in ``_process``.